File: src/trixi.f90
!>
!! @addtogroup api_f Fortran API
!!
!! **NOTE**: It is a known limitation of doxygen that Fortran interfaces will be listed as
!! "Data Types". Please refer to the "Functions/Subroutines" section instead.
!!
!! @{

module LibTrixi
  implicit none

  interface
    !>
    !! @fn LibTrixi::trixi_initialize_c::trixi_initialize_c(project_directory, depot_path)
    !!
    !! @brief Initialize Julia runtime environment (C char pointer version)
    !!
    !! Initialize Julia and activate the project at `project_directory`. If `depot_path` is
    !! given, forcefully set the environment variable `JULIA_DEPOT_PATH` to the value of
    !! depot_path`. If `depot_path` is not given, then proceed as follows: If
    !! `JULIA_DEPOT_PATH` is already set, do not touch it. Otherwise, set
    !! `JULIA_DEPOT_PATH` to `project_directory` + `default_depot_path`.
    !!
    !! @param[in]  project_directory  Path to project directory (C char pointer)
    !! @param[in]  depot_path         Path to Julia depot path (optional, C char pointer)
    !!
    !! @see @ref trixi_initialize        "trixi_initialize (Fortran convenience version)"
    !! @see @ref trixi_initialize_api_c  "trixi_initialize (C API)"
    subroutine trixi_initialize_c(project_directory, depot_path) bind(c, name='trixi_initialize')
      use, intrinsic :: iso_c_binding, only: c_char
      character(kind=c_char), dimension(*), intent(in) :: project_directory
      character(kind=c_char), dimension(*), intent(in), optional :: depot_path
    end subroutine

    !>
    !! @fn LibTrixi::trixi_initialize_simulation_c::trixi_initialize_simulation_c(libelexir)
    !!
    !! @brief Set up Trixi simulation (C char pointer version)
    !!
    !! @param[in]  libelixir  Path to libelexir file.
    !!
    !! @return handle (integer) to Trixi simulation instance
    !!
    !! @see @ref trixi_initialize_simulation       "trixi_initialize_simulation (Fortran convenience version)"
    !! @see @ref trixi_initialize_simulation_api_c "trixi_initialize_simulation (C API)"
    integer(c_int) function trixi_initialize_simulation_c(libelixir) bind(c, name='trixi_initialize_simulation')
      use, intrinsic :: iso_c_binding, only: c_char, c_int
      character(kind=c_char), dimension(*), intent(in) :: libelixir
    end function

    !>
    !! @fn LibTrixi::trixi_calculate_dt::trixi_calculate_dt(handle)
    !!
    !! @brief Get time step length
    !!
    !! @param[in]  handle  simulation handle
    !!
    !! @return Time step length
    !!
    !! @see @ref trixi_calculate_dt_api_c "trixi_calculate_dt (C API)"
    real(c_double) function trixi_calculate_dt(handle) bind(c)
      use, intrinsic :: iso_c_binding, only: c_int, c_double
      integer(c_int), value, intent(in) :: handle
    end function

    !>
    !! @fn LibTrixi::trixi_is_finished_c::trixi_is_finished_c(handle)
    !!
    !! @brief Check if simulation is finished (C integer version)
    !!
    !! @param[in]  handle  simulation handle
    !!
    !! @return 1 if finished, 0 if not
    !!
    !! @see @ref trixi_is_finished       "trixi_is_finished (Fortran convenience version)"
    !! @see @ref trixi_is_finished_api_c "trixi_is_finished (C API)"
    integer(c_int) function trixi_is_finished_c(handle) bind(c, name='trixi_is_finished')
      use, intrinsic :: iso_c_binding, only: c_int
      integer(c_int), value, intent(in) :: handle
    end function

    !>
    !! @fn LibTrixi::trixi_step::trixi_step(handle)
    !!
    !! @brief Perform next simulation step
    !!
    !! @param[in]  handle  simulation handle
    !!
    !! @see @ref trixi_step_api_c "trixi_step (C API)"
    subroutine trixi_step(handle) bind(c)
      use, intrinsic :: iso_c_binding, only: c_int
      integer(c_int), value, intent(in) :: handle
    end subroutine

    !>
    !! @fn LibTrixi::trixi_finalize_simulation::trixi_finalize_simulation(handle)
    !!
    !! @brief Finalize simulation
    !!
    !! @param[in]  handle  simulation handle
    !!
    !! @see trixi_finalize_simulation_api_c "trixi_finalize_simulation (C API)"
    subroutine trixi_finalize_simulation(handle) bind(c)
      use, intrinsic :: iso_c_binding, only: c_int
      integer(c_int), value, intent(in) :: handle
    end subroutine

    !>
    !! @fn LibTrixi::trixi_finalize::trixi_finalize()
    !!
    !! @brief Finalize Julia runtime environment.
    !!
    !! @see @ref trixi_finalize_api_c "trixi_finalize (C API)"
    subroutine trixi_finalize() bind(c)
    end subroutine

    !>
    !! @fn LibTrixi::julia_eval_string_c::julia_eval_string_c(code)
    !!
    !! @brief Execute Julia code (C char pointer version)
    !!
    !! @warning Only for development. Code is not checked prior to execution.
    !!
    !! @see @ref julia_eval_string       "julia_eval_string_c (Fortran convenience version)"
    !! @see @ref julia_eval_string_api_c "julia_eval_string_c (C API)"
    subroutine julia_eval_string_c(code) bind(c, name='julia_eval_string')
      use, intrinsic :: iso_c_binding, only: c_char
      character(kind=c_char), dimension(*), intent(in) :: code
    end subroutine

    !>
    !! @fn LibTrixi::trixi_version_major::trixi_version_major()
    !!
    !! @brief Return major version number of libtrixi.
    !!
    !! @return Major version number as integer.
    !!
    !! @see @ref trixi_version_major_api_c "trixi_version_major (C API)"
    integer(c_int) function trixi_version_major() bind(c)
      use, intrinsic :: iso_c_binding, only: c_int
    end function

    !>
    !! @fn LibTrixi::trixi_version_minor::trixi_version_minor()
    !!
    !! @brief Return minor version number of libtrixi.
    !!
    !! @return Minor version number as integer.
    !!
    !! @see @ref trixi_version_minor_api_c "trixi_version_minor (C API)"
    integer(c_int) function trixi_version_minor() bind(c)
      use, intrinsic :: iso_c_binding, only: c_int
    end function

    !>
    !! @fn LibTrixi::trixi_version_patch::trixi_version_patch()
    !!
    !! @brief Return patch version number of libtrixi.
    !!
    !! @return Patch version number as integer.
    !!
    !! @see @ref trixi_version_patch_api_c "trixi_version_patch (C API)"
    integer(c_int) function trixi_version_patch() bind(c)
      use, intrinsic :: iso_c_binding, only: c_int
    end function

    !>
    !! @fn LibTrixi::trixi_version_c::trixi_version_c()
    !!
    !! @brief Return full version string of libtrixi (C char pointer version).
    !!
    !! @return Full version string as C char pointer.
    !!
    !! @see @ref trixi_version       "trixi_version (Fortran convenience version)"
    !! @see @ref trixi_version_api_c "trixi_version (C API)"
    type(c_ptr) function trixi_version_c() bind(c, name='trixi_version')
      use, intrinsic :: iso_c_binding, only: c_ptr
    end function
  end interface

  contains

  !>
  !! @brief Check if simulation is finished (Fortran convenience version)
  !!
  !! @param[in]  handle  simulation handle
  !!
  !! @return true if finished, false if not
  !!
  !! @see @ref trixi_is_finished_c::trixi_is_finished_c
  !!           "trixi_is_finished (C integer version)"
  !! @see @ref trixi_is_finished_api_c
  !!           "trixi_is_finished (C API)"
  logical function trixi_is_finished(handle)
    use, intrinsic :: iso_c_binding, only: c_int
    integer(c_int), intent(in) :: handle

    trixi_is_finished = trixi_is_finished_c(handle) == 1
  end function

  !>
  !! @brief Initialize Julia runtime environment (Fortran convenience version)
  !!
  !! @param[in]  project_directory  Path to project directory (Fortran string).
  !! @param[in]  depot_path         Path to Julia depot path (Fortran string).
  !!
  !! @see @ref trixi_initialize_c::trixi_initialize_c
  !!           "trixi_initialize_c (C char pointer version)"
  !! @see @ref trixi_initialize_api_c
  !!           "trixi_initialize (C API)"
  subroutine trixi_initialize(project_directory, depot_path)
    use, intrinsic :: iso_c_binding, only: c_null_char
    character(len=*), intent(in) :: project_directory
    character(len=*), intent(in), optional :: depot_path

    if (present(depot_path)) then
      call trixi_initialize_c(trim(adjustl(project_directory)) // c_null_char, &
                              trim(adjustl(depot_path)) // c_null_char)
    else
      call trixi_initialize_c(trim(adjustl(project_directory)) // c_null_char)
    end if
  end subroutine

  !>
  !! @brief Set up Trixi simulation (Fortran convencience version)
  !!
  !! @param[in]  libelixir  Path to libelexir file.
  !!
  !! @return handle (integer) to Trixi simulation instance
  !!
  !! @see @ref trixi_initialize_simulation_c::trixi_initialize_simulation_c
  !!           "trixi_initialize_simulation_c (C char pointer version)"
  !! @see @ref trixi_initialize_simulation_api_c
  !!           "trixi_initialize_simulation (C API)"
  integer(c_int) function trixi_initialize_simulation(libelixir)
    use, intrinsic :: iso_c_binding, only: c_int, c_null_char
    character(len=*), intent(in) :: libelixir

    trixi_initialize_simulation = trixi_initialize_simulation_c(trim(adjustl(libelixir)) // c_null_char)
  end function

  !>
  !! @brief Execute Julia code (Fortran convenience version)
  !!
  !! @warning Only for development. Code is not checked prior to execution.
  !!
  !! @see @ref julia_eval_string_c::julia_eval_string_c
  !!           "julia_eval_string_c (C char pointer version)"
  !! @see @ref julia_eval_string_api_c
  !!           "julia_eval_string_c (C API)"
  subroutine julia_eval_string(code)
    use, intrinsic :: iso_c_binding, only: c_null_char
    character(len=*), intent(in) :: code

    call julia_eval_string_c(trim(adjustl(code)) // c_null_char)
  end subroutine

  !>
  !! @brief Return full version string of libtrixi (Fortran convenience version).
  !!
  !! @return Full version string as Fortran allocatable string.
  !!
  !! @see @ref trixi_version_c::trixi_version_c
  !!           "trixi_version_c (C char pointer version)"
  !! @see @ref trixi_version_api_c
  !!           "trixi_version (C API)"
  function trixi_version()
    use, intrinsic :: iso_c_binding, only: c_char, c_null_char, c_f_pointer
    character(len=:), allocatable :: trixi_version
    character(len=128, kind=c_char), pointer :: buffer
    integer :: length, i

    ! Associate buffer with C pointer
    call c_f_pointer(trixi_version_c(), buffer)

    ! Determine the actual length of the version string
    length = 0
    do i = 1,128
      if ( buffer(i:i) == c_null_char ) exit
      length = length + 1
    end do

    ! Store relevant part in return value
    trixi_version = buffer(1:(length + 1))
  end function
end module

!>
!! @}

File: src/trixi.c
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <julia.h>

#include "trixi.h"


// Auxiliary declarations for more helpful error messages
static void print_and_die(const char* message, const char* func, const char* file, int lineno);
#define LOC __func__, __FILE__, __LINE__

// Auxiliary function to evaluate julia REPL string with exception handling
static jl_value_t* checked_eval_string(const char* code, const char* func, const char* file, int lineno);

// Auxiliary function to determine debug level
static int show_debug_output();

// Auxiliary function to update JULIA_DEPOT_PATH environment variable
static void update_depot_path(const char * project_directory, const char * depot_path);

// Store function pointers to avoid overhead of `jl_eval_string`
enum {
  TRIXI_FTPR_INITIALIZE_SIMULATION,
  TRIXI_FTPR_CALCULATE_DT,
  TRIXI_FTPR_IS_FINISHED,
  TRIXI_FTPR_STEP,
  TRIXI_FTPR_FINALIZE_SIMULATION,
  TRIXI_FTPR_NDIMS,
  TRIXI_FTPR_NELEMENTS,
  TRIXI_FTPR_NVARIABLES,
  TRIXI_FTPR_LOAD_CELL_AVERAGES,
  TRIXI_FTPR_VERSION_JULIA,
  TRIXI_FTPR_VERSION_JULIA_EXTENDED,

  // The last one is for the array size
  TRIXI_NUM_FPTRS
};
static void* trixi_function_pointers[TRIXI_NUM_FPTRS];

// List of function names to obtain C function pointer from Julia
// OBS! If any name is longer than 250 characters, adjust buffer sizes below
static const char* trixi_function_pointer_names[] = {
  [TRIXI_FTPR_INITIALIZE_SIMULATION]  = "trixi_initialize_simulation_cfptr",
  [TRIXI_FTPR_CALCULATE_DT]           = "trixi_calculate_dt_cfptr",
  [TRIXI_FTPR_IS_FINISHED]            = "trixi_is_finished_cfptr",
  [TRIXI_FTPR_STEP]                   = "trixi_step_cfptr",
  [TRIXI_FTPR_FINALIZE_SIMULATION]    = "trixi_finalize_simulation_cfptr",
  [TRIXI_FTPR_NDIMS]                  = "trixi_ndims_cfptr",
  [TRIXI_FTPR_NELEMENTS]              = "trixi_nelements_cfptr",
  [TRIXI_FTPR_NVARIABLES]             = "trixi_nvariables_cfptr",
  [TRIXI_FTPR_LOAD_CELL_AVERAGES]     = "trixi_load_cell_averages_cfptr",
  [TRIXI_FTPR_VERSION_JULIA]          = "trixi_version_julia_cfptr",
  [TRIXI_FTPR_VERSION_JULIA_EXTENDED] = "trixi_version_julia_extended_cfptr"
};

// Default depot path *relative* to the project directory
// OBS! If you change the value here, you should also update the default value of
// `LIBTRIXI_JULIA_DEPOT` in `utils/libtrixi-init-julia` accordingly
static const char* default_depot_path = "julia-depot";


/**
 * @anchor trixi_initialize_api_c
 *
 * @brief Initialize Julia runtime environment
 * 
 * Initialize Julia and activate the project at `project_directory`. If `depot_path` is not
 * a null pointer, forcefully set the environment variable `JULIA_DEPOT_PATH` to the value
 * of `depot_path`. If `depot_path` *is* null, then proceed as follows:
 * If `JULIA_DEPOT_PATH` is already set, do not touch it. Otherwise, set `JULIA_DEPOT_PATH`
 * to `project_directory` + `default_depot_path`
 * 
 * @param[in]  project_directory  Path to project directory.
 * @param[in]  depot_path         Path to Julia depot path (optional; can be null pointer).
 */
void trixi_initialize(const char * project_directory, const char * depot_path) {
    // Update JULIA_DEPOT_PATH environment variable before initializing Julia
    update_depot_path(project_directory, depot_path);

    // Init Julia
    jl_init();

    // Construct activation command
    const char * activate = "using Pkg;\n"
                            "Pkg.activate(\"%s\"; io=devnull);\n";
    if ( strlen(activate) + strlen(project_directory) + 1 > 1024 ) {
      print_and_die("buffer size not sufficient for activation command", LOC);
    }
    char buffer[1024];
    snprintf(buffer, 1024, activate, project_directory);

    // Activate julia environment
    checked_eval_string(buffer, LOC);

    // Load LibTrixi module
    checked_eval_string("using LibTrixi;", LOC);
    if (show_debug_output()) {
      checked_eval_string("println(\"Module LibTrixi.jl loaded\")", LOC);
    }

    // Load function pointers
    char julia_command[256];
    for (int i = 0; i < TRIXI_NUM_FPTRS; i++) {
      // Reset for error detection
      trixi_function_pointers[i] = NULL;

      // Build Julia command
      snprintf(julia_command, 256, "%s()", trixi_function_pointer_names[i]);

      // Get and store function pointer
      trixi_function_pointers[i] = (void *)jl_unbox_voidpointer( checked_eval_string(julia_command, LOC) );

      // Perform sanity check
      if (trixi_function_pointers[i] == NULL) {
        fprintf(stderr, "ERROR: could not get function pointer with `%s()`\n",
                trixi_function_pointer_names[i]);
        print_and_die("null pointer", LOC);
      }
    }

    // Show version info
    if (show_debug_output()) {
      printf("\nLoaded julia packages:\n%s\n\n", trixi_version_julia());
    }
}


/**
 * @anchor trixi_initialize_simulation_api_c
 *
 * @brief Set up Trixi simulation
 *
 * Set up a Trixi simulation by reading the provided libelixir file. It resembles Trixi's
 * typical elixir files with the following differences:
 * - Everything (except `using ...`) has to be inside a `function init_simstate()`
 * - OrdinaryDiffEq's integrator has to be created by calling `init` (instead of `solve`)
 * - A `SimulationState` has to be created from the semidiscretization and the integrator
 * See the examples in the `LibTrixi.jl/examples` folder
 *
 * @param[in]  libelixir  Path to libelexir file.
 *
 * @return handle (integer) to Trixi simulation instance
 */
int trixi_initialize_simulation(const char * libelixir) {

    // Get function pointer
    int (*initialize_simulation)(const char *) = trixi_function_pointers[TRIXI_FTPR_INITIALIZE_SIMULATION];

    // Call function
    return initialize_simulation( libelixir );
}


/**
 * @anchor trixi_calculate_dt_api_c
 *
 * @brief Get time step length
 *
 * Get the current time step length of the simulation identified by handle.
 *
 * @param[in]  handle  simulation handle
 *
 * @return Time step length
 */
double trixi_calculate_dt(int handle) {

    // Get function pointer
    double (*calculate_dt)(int) = trixi_function_pointers[TRIXI_FTPR_CALCULATE_DT];;

    // Call function
    return calculate_dt( handle );
}


/**
 * @anchor trixi_is_finished_api_c
 *
 * @brief Check if simulation is finished
 *
 * Checks if the simulation identified by handle has reached its final time.
 *
 * @param[in]  handle  simulation handle
 *
 * @return 1 if finished, 0 if not
 */
int trixi_is_finished(int handle) {

    // Get function pointer
    int (*is_finished)(int) = trixi_function_pointers[TRIXI_FTPR_IS_FINISHED];

    // Call function
    return is_finished( handle );
}


/**
 * @anchor trixi_step_api_c
 *
 * @brief Perform next simulation step
 *
 * Let the simulation identified by handle advance by one step.
 *
 * @param[in]  handle  simulation handle
 */
void trixi_step(int handle) {

    // Get function pointer
    int (*step)(int) = trixi_function_pointers[TRIXI_FTPR_STEP];

    // Call function
    step( handle );
}


/**
 * @anchor trixi_finalize_simulation_api_c
 *
 * @brief Finalize simulation
 *
 * Finalize the simulation identified by handle. This will also release the handle.
 *
 * @param[in]  handle  simulation handle
 */
void trixi_finalize_simulation(int handle) {

    // Get function pointer
    void (*finalize_simulation)(int) = trixi_function_pointers[TRIXI_FTPR_FINALIZE_SIMULATION];

    // Call function
    finalize_simulation(handle);
}


/**
 * @anchor trixi_finalize_api_c
 *
 * @brief Finalize Julia runtime environment
 */
void trixi_finalize() {

    if (show_debug_output()) {
        printf("libtrixi: finalize\n");
    }

    // Reset function pointers
    for (int i = 0; i < TRIXI_NUM_FPTRS; i++) {
      trixi_function_pointers[i] = NULL;
    }

    jl_atexit_hook(0);
}


/**
 * @anchor trixi_ndims_api_c
 *
 * @brief Return number of spatial dimensions
 *
 * @param[in]  handle  simulation handle
 */
int trixi_ndims(int handle) {

    // Get function pointer
    int (*ndims)(int) = trixi_function_pointers[TRIXI_FTPR_NDIMS];

    // Call function
    return ndims(handle);
}


/**
 * @anchor trixi_nelements_api_c
 *
 * @brief Return number of elements (cells)
 *
 * @param[in]  handle  simulation handle
 */
int trixi_nelements(int handle) {

    // Get function pointer
    int (*nelements)(int) = trixi_function_pointers[TRIXI_FTPR_NELEMENTS];

    // Call function
    return nelements(handle);
}


/**
 * @anchor trixi_nvariables_api_c
 *
 * @brief Return number of (conservative) variables
 *
 * @param[in]  handle  simulation handle
 */
int trixi_nvariables(int handle) {

    // Get function pointer
    int (*nvariables)(int) = trixi_function_pointers[TRIXI_FTPR_NVARIABLES];

    // Call function
    return nvariables(handle);
}


/** 
 * @anchor trixi_load_cell_averages_api_c
 *
 * @brief Return cell averaged values
 *
 * Cell averaged values for each cell and each primitive variable are stored in a
 * contiguous array, where cell values for the first variable appear first and values for
 * the other variables subsequently (structure-of-arrays layout).
 *
 * The given array has to be of correct size and memory has to be allocated beforehand.
 *
 * @param[in]  handle  simulation handle
 * @param[out] data    cell averaged values for all cells and all primitive variables
 */
void trixi_load_cell_averages(double * data, int handle) {

    // Get function pointer
    void (*load_cell_averages)(double *, int) = trixi_function_pointers[TRIXI_FTPR_LOAD_CELL_AVERAGES];

    // Call function
    load_cell_averages(data, handle);
}


/**
 * @anchor trixi_version_julia_api_c
 *
 * @brief Return name and version of loaded julia packages LibTrixi directly depends on.
 *
 * The return value is a read-only pointer to a NULL-terminated string with the name and
 * version information of the loaded julia packages, separated by newlines.
 *
 * The returned pointer is to static memory and must not be used to change the contents of
 * the version string. Multiple calls to the function will return the same address.
 *
 * This function is thread-safe. It must be run after `trixi_initialize` has been called.
 *
 * @return Pointer to a read-only, NULL-terminated character array with the names and
 *         versions of loaded Julia packages.
 */
const char* trixi_version_julia() {

    // Get function pointer
    const char* (*version_julia)() = trixi_function_pointers[TRIXI_FTPR_VERSION_JULIA];

    // Call function
    return version_julia();
}


/**
 * @anchor trixi_version_julia_extended_api_c
 *
 * @brief Return name and version of all loaded julia packages.
 *
 * The return value is a read-only pointer to a NULL-terminated string with the name and
 * version information of all loaded julia packages, including implicit dependencies,
 * separated by newlines.
 *
 * The returned pointer is to static memory and must not be used to change the contents of
 * the version string. Multiple calls to the function will return the same address.
 *
 * This function is thread-safe. It must be run after `trixi_initialize` has been called.
 *
 * @return Pointer to a read-only, NULL-terminated character array with the names and
 *         versions of all loaded julia packages.
 */
const char* trixi_version_julia_extended() {

    // Get function pointer
    const char* (*version_julia_extended)() = trixi_function_pointers[TRIXI_FTPR_VERSION_JULIA_EXTENDED];

    // Call function
    return version_julia_extended();
}


/**
 * @anchor julia_eval_string_api_c
 *
 * @brief Execute Julia code
 *
 * Execute the provided code in the current Julia runtime environment.
 *
 * @warning Only for development. Code is not checked prior to execution.
 */
void julia_eval_string(const char * code) {

    checked_eval_string(code, LOC);
}



// Set JULIA_DEPOT_PATH environment variable appropriately
void update_depot_path(const char * project_directory, const char * depot_path) {
  // Set/modify Julia's depot path if desired
  if (depot_path != NULL) {
    // If depot path is provided as an argument, set environment variable JULIA_DEPOT_PATH
    // to it
    setenv("JULIA_DEPOT_PATH", depot_path, 1);
    if (show_debug_output()) {
      printf("JULIA_DEPOT_PATH set to \"%s\"\n", depot_path);
    }
  } else if (getenv("JULIA_DEPOT_PATH") == NULL) {
    // Otherwise, if environment variable is *not* already set, set it to
    // `project_directory` + `default_depot_path`

    // Verify that buffer size is large enough (+2 for '/' and trailing null)
    char path[1024];
    if ( strlen(project_directory) + strlen(default_depot_path) + 2 > 1024 ) {
      print_and_die("buffer size not sufficient for depot path construction", LOC);
    }

    // Construct complete path
    strcpy(path, project_directory);
    strcat(path, "/");
    strcat(path, default_depot_path);

    // Construct absolute path
    char absolute_path[PATH_MAX];
    realpath(path, absolute_path);

    // Set environment variable
    setenv("JULIA_DEPOT_PATH", absolute_path, 1);
    if (show_debug_output()) {
      printf("JULIA_DEPOT_PATH set to \"%s\"\n", absolute_path);
    }
  }
}


// Run Julia command and check for errors
// Source: https://github.com/JuliaLang/julia/blob/c0dd6ff8363f948237304821941b06d67014fa6a/test/embedding/embedding.c#L17-L31
jl_value_t* checked_eval_string(const char* code, const char* func, const char* file, int lineno) {

    jl_value_t *result = jl_eval_string(code);

    if (jl_exception_occurred()) {

        // none of these allocate, so a gc-root (JL_GC_PUSH) is not necessary
        jl_printf(jl_stderr_stream(), "ERROR in %s:%d (%s):\n", file, lineno, func);
        jl_printf(jl_stderr_stream(), "The following Julia code could not be evaluated: %s\n", code);
        jl_call2(jl_get_function(jl_base_module, "showerror"), jl_stderr_obj(), jl_exception_occurred());
        jl_printf(jl_stderr_stream(), "\n");
        jl_atexit_hook(1);
        exit(1);
    }

    assert(result && "Missing return value but no exception occurred!");

    return result;
}


void print_and_die(const char* message, const char* func, const char* file, int lineno) {
  fprintf(stderr, "ERROR in %s:%d (%s): %s\n", file, lineno, func, message);
  exit(1);
}


int show_debug_output() {
  const char * env = getenv("LIBTRIXI_DEBUG");
  if (env == NULL) {
    return 0;
  }

  if (strcmp(env, "all") == 0 || strcmp(env, "c") == 0) {
    return 1;
  } else {
    return 0;
  }
}
